Sync: Improves syncing save posts [In Progress]

lezama opened the pull request:
This PR intends to reduce the number of bytes that we sync on post save, by syncing a single clearer action.

Before this PR on saving a simple draft we were sending 4 different actions:
`set_object_terms` saving the categories
`set_object_terms` saving the tags
`jetpack_sync_save_post` saving the post
`jetpack_sync_save_post` (saving the revision, this one included again the title, content, excerpt, etc)

With this PR we Sync a single `jetpack_sync_save_post` object that includes as flags the `set_object_terms` details and everything we need to replicate the revision without sending duplicated values.


Things to test:

- [x] Publish post
- [x] Edit post
- [x] Delete Post
- [x] Trash
- [x] Publicize
- [x] Subscriptions
- [x] Rewind restore point triggered
- [x] Bulk updateing (quick edit UI in wp-admin)
- [x] Import posts.

lezama committed:
testing ideas

lezama committed:
maybe?

lezama committed:
Add terms and revision to the flags

lezama committed:
unset duplicated revsion content

lezama committed:
remove error_log

roccotripaldi reviewed (commented):
@gititon and I ran through this a few times, and it seems pretty cool! We noticed that it is adding object terms to the sync object for posts, instead of having separate events.

There is some strangeness that happens when you first click "new post" - it seems like in that case we have object terms. I'll see if i can fix that. 

Also, i'd love to see a description for this PR that explains why we are doing this. Is it to reduce the amount of database writes?

enejb committed:
Fix some of the tests.

Many more still failing

enejb committed:
Fix more tests

enejb committed:
New experimental Jetpack Sync Item object

Note that is there is no post_id is passed that we don't capture the expected events for the taxonomy.

enejb commented:
cc: @lezama and @gititon What do you think? 

To do here
- Add tests that show the taxonomy events being captured. 
- Should we also include the post_meta changes being made. 
- Add test for this as well. 
- Make the updates on the .com side of things. 
- Make the test pass here

enejb committed:
Fix minor test

lezama commented:
let's do the post meta on a second PR as on a simple post saving there's no post meta being synced.

lezama commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+class Jetpack_Sync_Item {
+
+	private $object;
+	private $sync_items;
```
I think we should add a comment about what we intend to store on each attribute

lezama commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+
+	function __construct( $trigger, $object = null ) {
+		$this->trigger = $trigger;
+		if ( !is_null( $object ) ) {
```
missing space after `!`

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -55,75 +59,42 @@ public function init_listeners( $callable ) {
 		add_action( 'jetpack_full_sync_posts', $callable ); // also sends post meta
 	}
 
 	public function init_before_send() {
```
In the next phase of refactoring, I'd love to see the sender call a `process_before_send()` method on each sync item, where each sync item is a subclass containing the code necessary to do whatever processing it needs to, rather than this mechanism.

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
 
-		$this->previous_status[ $post->ID ] = $old_status;
+
+		if ( ! isset( $this->sync_items[ $post->ID ] ) ) {
```
I think we can make the code even more accessible if we make an alias method that describes what we're actually testing for here. That is, instead of `if ( ! isset( $this->sync_items[ $post->ID ] ) ) {`, something like `if ( $this->is_new_post( $post->ID ) ) {`

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
+			$this->sync_items[ $post->ID ] = new Jetpack_Sync_Item( 'save_post' );
+		}
+		$sync_item = $this->sync_items[ $post->ID ];
+		$just_published = 'publish' === $new_status && 'publish' !== $old_status;
```
Consider using variable names that start with `is_` to make it clear that they hold boolean values, e.g. `$is_just_published`

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
+		}
+		$sync_item = $this->sync_items[ $post->ID ];
+		$just_published = 'publish' === $new_status && 'publish' !== $old_status;
+		$sync_item->add_meta( 'just_published', $just_published );
```
Consider using key names that start with is_ to make it clear that they hold boolean values, e.g. `is_just_published`.

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
+			unset( $post['post_content'] );
+			unset( $post['post_title'] );
+			unset( $post['post_excerpt'] );
+			$sync_item = $this->sync_items[ $post['post_parent']];
```
Need space before closing square bracket

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
 			return;
 		}
 
+		if ( wp_is_post_revision( $post ) && $this->is_saving_post( $post->post_parent ) ) {
```
Consider replacing this block of code with an alias function and a method to do the work, just to make things more orderly and compartmentalize them:
```
if ( $this->is_revision( $post ) {
  $this->process_revision( $post );
  return;
}
```

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
-	public function send_published( $post_ID, $post ) {
-		if ( ! isset( $this->just_published[ $post_ID ] ) ) {
-			return;
+		if ( ! isset( $this->sync_items[ $post_ID ] ) ) {
```
I think we can make the code even more accessible if we make an alias method that describes what we're actually testing for here. That is, instead of `if ( ! isset( $this->sync_items[ $post->ID ] ) ) {`, something like `if ( $this->is_new_post( $post->ID ) ) {`

gititon replied:
I agree

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
-		$post_flags = array(
-			'post_type' => $post->post_type
-		);
+		$sync_item->add_meta('is_auto_save', (bool) Jetpack_Constants::get_constant( 'DOING_AUTOSAVE' ) );
```
Do we need a space after `(` before `'` ?

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
-			'post_type' => $post->post_type
-		);
+		$sync_item->add_meta('is_auto_save', (bool) Jetpack_Constants::get_constant( 'DOING_AUTOSAVE' ) );
+		$sync_item->add_meta('update', $update );
```
Do we need a space after `(` before `'` ?

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -316,76 +286,84 @@ function filter_post_content_and_add_links( $post_object ) {
+			 *
+			 * @since 5.9.0
+			 *
+			 * @param mixed $post WP_POST object
```
Are these params still accurate?

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -423,4 +400,68 @@ public function expand_post_ids( $args ) {
+	/**
+	 * IMPORT SECTIONS .. figures out the whole importing of things.
+	 */
+	public function sync_import_done( $importer ) {
```
It doesn't have to be in this PR, but lets see if we can get this import business into a file/module of its own in the near future.

gititon reviewed (changes requested):
Looking good! Just a few suggestions.

roccotripaldi committed:
Do not use `func_get_args` as a param, because it is causing fatal errors.

roccotripaldi committed:
Cleaning up spacing, and address feedback from @gititon

roccotripaldi committed:
- renaming `meta_data` to `state`, because `meta` already has a specific meaning in regards to WordPress posts
- adding comments and doc blocks

roccotripaldi committed:
- update unit tests

gititon commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+
+	function __construct( $trigger, $object = null ) {
+		$this->trigger = $trigger;
+		if ( ! is_null( $object ) ) {
```
Why not `if ( object ) { }` ?

gititon commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	function add_terms( Jetpack_Sync_Item $terms ) {
+		$this->terms[] = $terms;
```
Will we always have one batch of terms to add to a sync object? Would there be any reason to array_merge the array of terms that are passed as an argument with the existing array of terms?

roccotripaldi replied:
This is consistent with how we've set object terms in the past, but i agree that it should be more clear. we can address that in an other PR

gititon commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+		return $this->object;
+	}
+
+	function set_state_value( $key, $value = null ) {
```
When would we want to set a key with a null value?

gititon commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	function is_state_value_true( $key ) {
+		return (bool) ( $this->state_isset( $key ) && (bool) $this->state[ $key ] );
```
You can remove the cast to `(bool)` beside return, since `$this->state_isset( $key )` necessarily resolves to a bool, and `(bool) $this->state[ $key ]` necessarily resolves to a bool, therefore &&'ing them together will necessarily resolve to a bool.

gititon commented on sync/class.jetpack-sync-item.php:
```diff
@@ -0,0 +1,79 @@
+
+	private function add( &$array, $key, $value = null ) {
+		if ( is_array( $key ) ) {
+			$array = array_merge( $key, $array );
```
I second the above

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -55,75 +59,41 @@ public function init_listeners( $callable ) {
 
 	public function init_before_send() {
-		add_filter( 'jetpack_sync_before_send_jetpack_sync_save_post', array( $this, 'expand_jetpack_sync_save_post' ) );
+		add_filter( 'jetpack_sync_before_send_jetpack_post_saved', array( $this, 'expand_jetpack_post_saved' ) );
```
If we want to go further with the transition, we could put `expand_jetpack_post_saved` and `$this->filter_post_content_and_add_links` on a subclass of the sync item object, and simply reference the sync item object's `expand_jetpack_post_saved` method here, instead.

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -318,98 +287,111 @@ function filter_post_content_and_add_links( $post_object ) {
+		$sync_item = $this->sync_items[ $post->ID ];
+		$is_just_published = 'publish' === $new_status && 'publish' !== $old_status;
+		$sync_item->set_state_value( 'is_just_published', $is_just_published );
+		$sync_item->set_state_value( 'previous_status', $old_status );
```
Should we consider having explicit get_state(), set_state(), get_previous_state(), and set_previous_state() methods on the sync object class?

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -318,98 +287,111 @@ function filter_post_content_and_add_links( $post_object ) {
-		// Post revisions cause race conditions where this send_published add the action before the actual post gets synced
-		if ( wp_is_post_autosave( $post ) || wp_is_post_revision( $post ) ) {
-			return;
+		if ( ! $sync_item->state_isset( 'previous_status' ) ) {
```
Would be prettier as `if ( ! $sync_item->get_previous_state() ) { }` or `if ( is_null( ... ) )`

gititon commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -318,98 +287,111 @@ function filter_post_content_and_add_links( $post_object ) {
-		if ( wp_is_post_autosave( $post ) || wp_is_post_revision( $post ) ) {
-			return;
+		if ( ! $sync_item->state_isset( 'previous_status' ) ) {
+			$sync_item->set_state_value( 'previous_status', self::DEFAULT_PREVIOUS_STATE );
```
Would be prettier as `$sync_item->set_previous_state( ... )`

gititon replied:
Although maybe not unless previous_state is something many sync objects are going to have/use.

roccotripaldi committed:
Addressing feedback from @gititon

roccotripaldi committed:
Update AMP unit test to use new schema

roccotripaldi committed:
`is_saving_post` should account for 'new' as a sync item key

roccotripaldi committed:
Improve clarity of setting sync items.

roccotripaldi committed:
No need to use concept of 'new'

lezama commented:
We decided that we are going to merge this one after the freeze, as it needs deep testing :)

roccotripaldi committed:
Further "improvements" to set_object_terms

roccotripaldi committed:
Ensure we unset new sync items once we have indentified them.

roccotripaldi committed:
fixing post unit tests

roccotripaldi commented:
Importing posts results in a `set_object_terms` action. i'll address that in an other PR.

enejb committed:
Add a more generic way to add sync items

enejb committed:
Add support for post meta

roccotripaldi commented:
Ooof - started testing this against Menus and Customizer today, and I don't think it is working as expected.

lezama committed:
Simplifying get_payload

roccotripaldi committed:
fix some of the tests

roccotripaldi committed:
Ensure we don't initialize post sync items for attachments

roccotripaldi commented:
Additional server-side changes merged in 172966-r

roccotripaldi committed:
Improving logic around post sync sub items

roccotripaldi commented:
I think it is ready to go!

oskosk commented:
Looks like there's some feedback to be addressed . Can someone approve this one ?

zinigor commented:
@gititon can you take a look at this again please?

jeherve commented:
Closing this for now.

lezama commented:
@Automattic/poseidon @gravityrail  what do you think about this one? should we rebase and try to get it merged?

gravityrail commented:
What are the exact savings we could expect to see? 

How much work is required to get this into a mergeable state?
